src/generators/modelsGenerator.ts

```diff
@@ -16,31 +37,332 @@ export function generateModels(clientDetails: ClientDetails, project: Project) {
+  modelsIndexFile: SourceFile
+) => clientDetails.objects.forEach(writeObjectSignature(modelsIndexFile));
+
+const writeObjectSignature = (modelsIndexFile: SourceFile) => (
+  model: ObjectDetails
+) => {
+  const properties = getPropertiesSignatures(model);
+  const parents = model.parents.map(p => p.name).join(" & ");
+
+  if (parents) {
+    modelsIndexFile.addTypeAlias({
+      name: model.name,
+      docs: [model.description],
+      isExported: true,
+      type: Writers.intersectionType(
+        parents,
+        Writers.objectType({ properties })
+      ),
+      leadingTrivia: writer => writer.blankLine()
+    });
+  } else {
     modelsIndexFile.addInterface({
       name: model.name,
       docs: [model.description],
       isExported: true,
-      properties: model.properties
-        .filter(p => !p.isConstant)
-        .map<PropertySignatureStructure>(p => ({
-          name: p.name,
-          hasQuestionToken: !p.required,
-          isReadonly: p.readOnly,
-          type: p.type,
-          docs: p.description ? [p.description] : undefined,
-          kind: StructureKind.PropertySignature
-        }))
+      properties,
+      leadingTrivia: writer => writer.blankLine()
     });
   }
+};
 
-  for (const union of clientDetails.unions) {
-    modelsIndexFile.addTypeAlias({
-      name: union.name,
-      docs: [union.description],
-      isExported: true,
-      type: union.values.join(" | "),
-      trailingTrivia: writer => writer.newLine()
+/**
+ * This function writes all UnionTypes, these types represent the optios a request can use for a Polymorphic parameter
+ */
+function writeUniontypes({ objects }: ClientDetails, modelsFile: SourceFile) {
+  objects
+    .filter(
+      obj => obj.kind === ObjectKind.Polymorphic && obj.children.length > 0
+    )
+    .forEach(obj => {
+      const polymorphicObject = obj as PolymorphicObjectDetails;
+      const childrenNames = [
+        obj.name,
+        ...polymorphicObject.children.map(c => {
+          return c.schema.children && c.schema.children.immediate.length
+            ? `${c.name}Union`
+            : c.name;
+        })
+      ];
+      modelsFile.addTypeAlias({
+        name: `${obj.name}Union`,
+        isExported: true,
+        type: childrenNames.join(" | "),
+        trailingTrivia: writer => writer.newLine()
+      });
     });
+}
+
+function writeOptionalParameters(
+  operationGroupName: string,
+  operationName: string,
+  optionalParams: ParameterDetails[],
+  modelsIndexFile: SourceFile
+) {
+  if (!optionalParams || !optionalParams.length) {
+    return;
   }
+
+  modelsIndexFile.addInterface({
+    name: `${operationGroupName}${operationName}OptionalParams`,
+    docs: ["Optional parameters."],
+    isExported: true,
+    extends: ["coreHttp.RequestOptionsBase"],
+    properties: optionalParams.map<PropertySignatureStructure>(p => ({
+      name: p.name,
+      hasQuestionToken: !p.required,
+      type: p.typeDetails.typeName || "TODO NO TYPE",

```

Comment:
Is this always going to be an error case or just something we need to track adding later?

Reply:
Forgot to remove this. it should be handled when populating typeDetails